cs/src/core/Index/FasterLog/FasterLog.cs

```diff
@@ -0,0 +1,472 @@
+                logSettings.LogDevice.FileName + ".commit");
+
+            epoch = new LightEpoch();
+            CommittedUntilAddress = Constants.kFirstValidAddress;
+            allocator = new BlittableAllocator<Empty, byte>(
+                logSettings.GetLogSettings(), null, 
+                null, epoch, e => Commit(e));
+            allocator.Initialize();
+            Restore();
+        }
+
+        /// <summary>
+        /// Dispose
+        /// </summary>
+        public void Dispose()
+        {
+            allocator.Dispose();
+            epoch.Dispose();
+        }
+
+        /// <summary>
+        /// Append entry to log
+        /// </summary>
+        /// <param name="entry"></param>
+        /// <returns>Logical address of added entry</returns>
+        public long Append(Span<byte> entry)
+        {
+            long logicalAddress;
+            while (!TryAppend(entry, out logicalAddress)) ;
+            return logicalAddress;
+        }
+
+        /// <summary>
+        /// Append entry to log
+        /// </summary>
+        /// <param name="entry"></param>
+        /// <returns>Logical address of added entry</returns>
+        public long Append(byte[] entry)
+        {
+            long logicalAddress;
+            while (!TryAppend(entry, out logicalAddress)) ;
+            return logicalAddress;
+        }
+
+        /// <summary>
+        /// Try to append entry to log. If it returns true, we are
+        /// done. If it returns false, we need to retry.
+        /// </summary>
+        /// <param name="entry">Entry to be appended to log</param>
+        /// <param name="logicalAddress">Logical address of added entry</param>
+        /// <returns>Whether the append succeeded</returns>
+        public unsafe bool TryAppend(byte[] entry, out long logicalAddress)
+        {
+            logicalAddress = 0;
+
+            epoch.Resume();
+
+            var length = entry.Length;
+            logicalAddress = allocator.TryAllocate(4 + Align(length));
+            if (logicalAddress == 0)
+            {
+                epoch.Suspend();
+                return false;
+            }
+
+            var physicalAddress = allocator.GetPhysicalAddress(logicalAddress);
+            *(int*)physicalAddress = length;
+            fixed (byte* bp = entry)
+                Buffer.MemoryCopy(bp, (void*)(4 + physicalAddress), length, length);
+
+            epoch.Suspend();
+            return true;
+        }
+
+        /// <summary>
+        /// Try to append entry to log. If it returns true, we are
+        /// done. If it returns false, we need to retry.
+        /// </summary>
+        /// <param name="entry">Entry to be appended to log</param>
+        /// <param name="logicalAddress">Logical address of added entry</param>
+        /// <returns>Whether the append succeeded</returns>
+        public unsafe bool TryAppend(Span<byte> entry, out long logicalAddress)
```

Comment:
Nit: You could make this a `ReadOnlySpan` to communicate semantics a bit better

Reply:
Done.